### ISSUE TITLE
DFS network splitting function hits a snag when used with reservoirs

### ISSUE THREAD

**Issue by jameshalgren** [NOAA-OWP/t-route#343]
The `nhd_network.dfs_decomposition_depth_tuple` function, when applied in the up-and-coming diffusive-with-reservoirs application uses `split_at_junctions` as the `path_func` but should use `split_at_waterbodies_and_junctions` as the `path_func`. With `split_at_junctions`, the function returns an order of negative 1 for a network consisting of a single segment with an upstream that is a reservoir. 

e.g.
```
> from functools import partial
> rconn = {"down":["up (is a reservoir)"]}
> path_func = partial(nhd_network.split_at_junction, rconn)
> nhd_network.dfs_decomposition_depth_tuple(rconn, path_func)
[(-1, ['down'])]
```
While using what should be the appropriate splitting function yields an error:
```
> path_func = partial(nhd_network.split_at_waterbodies_and_junctions, ["up (is a reservoir)"], rconn)
> nhd_network.dfs_decomposition_depth_tuple(rconn, path_func)
*** TypeError: 'NoneType' object is not subscriptable
``` 

What is needed is probably a small mod to that function or pair of functions. 

In the regular dfs function `dfs_decomposition`, the splitting function (applied as a partial, `path_func`) is used only on the reverse traversal, i.e., after the depth-first search has reached a leaf of the tree. As it pops items out of the stack, the path variable keeps track of the values coming down the line.
As this happens, the splitting function examines both the node and the path. If the path is a reservoir and the node is not OR if the node is a reservoir and the path is not (i.e., if we are entering or leaving a reservoir), then the function is triggered and the path splits.

The problem is that in the slightly more complex `dfs_decomposition_depth_tuple`, I am trying to use that same splitting function in reverse on the way up the tree to increment a count of the reach depth... [this code](https://github.com/NOAA-OWP/t-route/blob/master/src/python_framework_v02/troute/nhd_network.py#L246-L248).

There is not a "path" defined at that point. So I was passing None to the function -- which works with the junction only splitting function (which doesn't even reference the path argument...) -- but the reservoir splitting function needs the path to detect the change.

_Possible Solutions_
* The `(order, reach)` tuples returned from the decomposition could be made by coalescing the reaches (e.g., in the `reaches_bytw` object and counting the depth. 
* The if statement could be rebuilt to accumulate the logic from the collection of potential children:
```
is_new_split = False
for _child in N[node]:  # list of upstreams
    if not path_func([node], _child):
        is_new_split = True  # if any of the children are a split, then they all should be.
        break
if is_new_split:
    if node in junctions:
        is_new_split = False
```

**Comment by jameshalgren** [comment]
@awlostowski-noaa @jdmattern-noaa 
The logic I've given above seems to work for the isolated test case and, if I'm not mistaken, it does so for the right reason. I just submitted #344 as a potential fix for this.

**Comment by awlostowski-noaa** [comment]
@jameshalgren Thanks for the detailed issue, here. This should address exactly what was causing issues in our session earlier this afternoon.

**Comment by jdmattern-noaa** [comment]
@jameshalgren Thanks. I will add this to my implementation also and continue forth with the reservoir boundary conditions.

**Comment by jameshalgren** [comment]
Issued addressed first in proposed/closed #344; finally address in now-merged #347
